sql/src/main/java/io/crate/execution/engine/join/JoinOperations.java

```diff
@@ -113,4 +125,93 @@ public static Projection createJoinProjection(List<Symbol> outputs, List<Symbol>
+        LongBitSet processedJoinPairs = new LongBitSet(explicitJoinPairs.size());
+        Iterator<Map.Entry<Set<QualifiedName>, Symbol>> queryIterator = splitQueries.entrySet().iterator();
+        List<JoinPair> newJoinPairs = new ArrayList<>(explicitJoinPairs.size() + splitQueries.size());
+
+        while (queryIterator.hasNext()) {
+            Map.Entry<Set<QualifiedName>, Symbol> queryEntry = queryIterator.next();
+            Set<QualifiedName> relations = queryEntry.getKey();
+
+            if (relations.size() == 2) { // If more than 2 relations are involved it cannot be converted to a JoinPair
+                Symbol implicitJoinCondition = queryEntry.getValue();
+                JoinPair newJoinPair = null;
+                for (int i = 0; i < explicitJoinPairs.size(); i++) {
+                    JoinPair joinPair = explicitJoinPairs.get(i);
+                    if (relations.contains(joinPair.left()) && relations.contains(joinPair.right())) {
+                        processedJoinPairs.set(i);
+                        // If a JoinPair with the involved relations already exists then depending on the JoinType:
+                        //  - INNER JOIN: the implicit join condition can be "AND joined" with
+                        //                the existing explicit condition of the JoinPair.
+                        //
+                        //  - CROSS JOIN: the implicit join condition becomes explicit and set on the JoinPair
+                        //                which becomes an INNER JOIN
+                        //
+                        //  - Outer Joins: Queries in the WHERE clause must be semantically applied after the outer join
+                        //                 and cannot be merge with the conditions of the ON clause.
+                        //
+                        if (joinPair.joinType() == JoinType.INNER || joinPair.joinType() == JoinType.CROSS) {
+                            joinPair.joinType(JoinType.INNER);
+                            newJoinPair = JoinPair.of(
+                                joinPair.left(),
+                                joinPair.right(),
+                                JoinType.INNER,
+                                mergeJoinConditions(joinPair.condition(), implicitJoinCondition));
+                            queryIterator.remove();
+                        } else {
+                            newJoinPair = joinPair;
+                        }
+                    }
+                }
+                if (newJoinPair == null) {
+                    Iterator<QualifiedName> namesIter = relations.iterator();
+                    newJoinPair = JoinPair.of(namesIter.next(), namesIter.next(), JoinType.INNER, implicitJoinCondition);
+                    queryIterator.remove();
+                }
+
+                newJoinPairs.add(newJoinPair);
+            }
+        }
+
+        // Add not-processed existing JoinPairs
```

Comment:
Might want to extract this into a method and remove the comment. (Avoiding "what" comments by using naming/functions)